File: node/src/humanize.js
import fs from 'node:fs/promises';
import path from 'node:path';
import { CLIError } from './client.js';
import { t } from './i18n.js';

const RANDOM_FIELDS = ['avatar', 'name', 'bio', 'pronouns', 'hypesquad'];
const HUMANIZE_FIELDS = new Set(['avatar', 'banner', 'name', 'bio', 'pronouns', 'hypesquad']);
const HYPESQUAD = new Map([['1','1'], ['bravery','1'], ['2','2'], ['brilliance','2'], ['3','3'], ['balance','3']]);
const MIME_BY_EXT = new Map([['.png','image/png'], ['.jpg','image/jpeg'], ['.jpeg','image/jpeg'], ['.gif','image/gif'], ['.webp','image/webp']]);

export function normalizeHypesquad(value) {
  const normalized = HYPESQUAD.get(String(value).trim().toLowerCase());
  if (!normalized) throw new CLIError(t('hypesquadInvalid'));
  return normalized;
}

export function validateHumanize(config) {
  if (!config || typeof config !== 'object' || Array.isArray(config) || Object.keys(config).length === 0) throw new CLIError(t('humanizeRequired'));
  const unknown = Object.keys(config).filter((field) => !HUMANIZE_FIELDS.has(field));
  if (unknown.length) throw new CLIError(t('humanizeUnknown', { fields: unknown.sort().join(', ') }));
  for (const [field, spec] of Object.entries(config)) {
    if (!spec || typeof spec !== 'object' || Array.isArray(spec)) throw new CLIError(t('humanizeSpec', { field }));
    if (!['random', 'custom'].includes(spec.source)) throw new CLIError(t('humanizeSource', { field }));
    if (field === 'banner' && spec.source === 'random') throw new CLIError(t('bannerCustom'));
    if (spec.source === 'custom' && !Object.hasOwn(spec, 'value')) throw new CLIError(t('customValue', { field }));
    const custom = spec.source === 'custom' ? spec.value : undefined;
    if (field === 'name' && typeof custom === 'string' && [...custom].length > 32) throw new CLIError(t('nameLength'));
    if (field === 'bio' && typeof custom === 'string' && [...custom].length > 190) throw new CLIError(t('bioLength'));
    if (field === 'pronouns' && typeof custom === 'string' && [...custom].length > 40) throw new CLIError(t('pronounsLength'));
    if (field === 'hypesquad' && spec.source === 'custom' && !['1','2','3'].includes(String(custom))) throw new CLIError(t('hypesquadInvalid'));
  }
  return config;
}

async function readText(file) {
  try { return await fs.readFile(file, 'utf8'); }
  catch { throw new CLIError(t('readFile', { path: file })); }
}

export async function imageFileToDataUrl(file) {
  const ext = path.extname(file).toLowerCase();
  const mime = MIME_BY_EXT.get(ext);
  if (!mime) throw new CLIError(t('imageType', { path: file }));
  let data;
  try { data = await fs.readFile(file); } catch { throw new CLIError(t('imageNotFound', { path: file })); }
  return `data:${mime};base64,${data.toString('base64')}`;
}

async function loadLegacy(value) {
  if (!value) return {};
  const raw = value.startsWith('@') ? await readText(value.slice(1)) : value;
  let parsed;
  try { parsed = JSON.parse(raw); } catch (error) { throw new CLIError(t('jsonInvalid', { error: error.message })); }
  return validateHumanize(parsed);
}

export async function buildHumanizeConfig(values, { required = false } = {}) {
  const config = { ...(await loadLegacy(values.humanizeJson)) };
  if (values.randomAll) for (const field of RANDOM_FIELDS) config[field] = { source: 'random' };

  if (values.randomAvatar) config.avatar = { source: 'random' };
  else if (values.avatarUrl) {
    const value = String(values.avatarUrl).trim();
    if (!/^(https?:\/\/|data:image\/)/i.test(value)) throw new CLIError(t('avatarInvalid'));
    config.avatar = { source: 'custom', value };
  } else if (values.avatarFile) config.avatar = { source: 'custom', value: await imageFileToDataUrl(values.avatarFile) };

  if (values.bannerFile) config.banner = { source: 'custom', value: await imageFileToDataUrl(values.bannerFile) };
  else if (values.bannerData) {
    const value = String(values.bannerData).trim();
    if (!/^data:image\//i.test(value)) throw new CLIError(t('bannerInvalid'));
    config.banner = { source: 'custom', value };
  }

  for (const field of ['name', 'bio', 'pronouns']) {
    const cap = field[0].toUpperCase() + field.slice(1);
    if (values[`random${cap}`]) config[field] = { source: 'random' };
    else if (values[field] !== undefined) config[field] = { source: 'custom', value: values[field] };
  }
  if (values.randomHypesquad) config.hypesquad = { source: 'random' };
  else if (values.hypesquad !== undefined) config.hypesquad = { source: 'custom', value: normalizeHypesquad(values.hypesquad) };

  if (Object.keys(config).length === 0) {
    if (required) throw new CLIError(t('humanizeRequired'));
    return undefined;
  }
  return validateHumanize(config);
}

File: src/salta7_cli/humanize.py
from __future__ import annotations

import base64
import mimetypes
from pathlib import Path
from typing import Any, Mapping, Optional

from .client import CLIError
from .i18n import t
from .utils import load_json_value, validate_humanize

RANDOM_FIELDS = ("avatar", "name", "bio", "pronouns", "hypesquad")
HYPESQUAD_ALIASES = {
    "1": "1",
    "bravery": "1",
    "2": "2",
    "brilliance": "2",
    "3": "3",
    "balance": "3",
}


def image_file_to_data_url(path: str) -> str:
    file_path = Path(path).expanduser()
    if not file_path.is_file():
        raise CLIError(t("humanize.image_not_found", path=file_path))
    mime, _ = mimetypes.guess_type(file_path.name)
    if not mime or not mime.startswith("image/"):
        raise CLIError(t("humanize.image_type", path=file_path))
    encoded = base64.b64encode(file_path.read_bytes()).decode("ascii")
    return f"data:{mime};base64,{encoded}"


def normalize_hypesquad(value: str) -> str:
    normalized = HYPESQUAD_ALIASES.get(str(value).strip().lower())
    if normalized is None:
        raise CLIError(t("utils.hypesquad_value"))
    return normalized


def build_humanize_config(values: Mapping[str, Any], *, required: bool = False) -> Optional[dict[str, Any]]:
    """Build the API Humanize object from CLI-friendly options.

    Legacy --humanize-json is used as a base. Explicit CLI flags override the
    matching fields, while --random-all fills every random-capable field.
    """

    config = dict(load_json_value(values.get("humanize_json")) or {})

    if values.get("random_all"):
        for field in RANDOM_FIELDS:
            config[field] = {"source": "random"}

    if values.get("random_avatar"):
        config["avatar"] = {"source": "random"}
    elif values.get("avatar_url"):
        avatar_url = str(values["avatar_url"]).strip()
        if not avatar_url.startswith(("http://", "https://", "data:image/")):
            raise CLIError(t("humanize.avatar_url_invalid"))
        config["avatar"] = {"source": "custom", "value": avatar_url}
    elif values.get("avatar_file"):
        config["avatar"] = {"source": "custom", "value": image_file_to_data_url(values["avatar_file"])}

    if values.get("banner_file"):
        config["banner"] = {"source": "custom", "value": image_file_to_data_url(values["banner_file"])}
    elif values.get("banner_data"):
        banner_data = str(values["banner_data"]).strip()
        if not banner_data.startswith("data:image/"):
            raise CLIError(t("humanize.banner_data_invalid"))
        config["banner"] = {"source": "custom", "value": banner_data}

    for field in ("name", "bio", "pronouns"):
        if values.get(f"random_{field}"):
            config[field] = {"source": "random"}
        elif values.get(field) is not None:
            config[field] = {"source": "custom", "value": values[field]}

    if values.get("random_hypesquad"):
        config["hypesquad"] = {"source": "random"}
    elif values.get("hypesquad") is not None:
        config["hypesquad"] = {"source": "custom", "value": normalize_hypesquad(values["hypesquad"])}

    if not config:
        if required:
            raise CLIError(t("humanize.required"))
        return None

    validate_humanize(config)
    return config
